feat(diff): add word-level highlighting for modified lines

remorses opened the pull request:
Adds word-level highlighting to show which words changed within modified lines (like GitHub's diff view).

**Algorithm:**
1. Parse diff and group consecutive remove/add lines into change blocks
2. Pair lines positionally within each block (1st remove with 1st add, etc.)
3. For each pair, compute similarity score using character-level diff
4. If similarity >= 40%, compute word boundaries with `diffWordsWithSpace`
5. Pass highlight column ranges to LineNumberRenderable for rendering

Complexity is O(L + B×P×C²) where L=lines, B=change blocks, P=pairs per block, C=chars per line. In practice C is small (~100) and only change blocks are processed.

**Performance guards:** skips word diff for dissimilar lines, skips entirely for large blocks (>50 lines).

<img width="1702" height="850" alt="Screenshot 2025-11-30 at 18 01 50" src="https://github.com/user-attachments/assets/45c01136-21bb-483f-8931-fe431efdac33" />

**New options:**
- `disableWordHighlights` - Turn off word-level highlighting (default: `false`)
- `addedWordBg` - Background color for added words (default: brighter `addedBg`)
- `removedWordBg` - Background color for removed words (default: brighter `removedBg`)
- `lineSimilarityThreshold` - Min similarity (0-1) to pair lines for word diff (default: `0.4`)

**Also includes:**
- Fix for word highlights scrolling with content when `wrapMode: "none"` and horizontal scrolling is used

remorses committed:
add word highlights

remorses committed:
test: add word highlight tests

remorses committed:
format

kommander commented:
This would be cool. I'd make it an option though so both view styles work. 

I am concerned about multi width graphemes though, it cannot use String.length or utf16 char offsets from js as the highlights operate on display width currently.

That's a bit tricky. What could work to get content start and end for the highlight offsets would be to measure the string before the content starts with Bun.stringWidth, then the content that should be highlighted itself with Bun.stringWidth and add that to the former for the end offset.

remorses committed:
fix: use Bun.stringWidth for multi-width character support in word highlights

remorses committed:
fix: correct line counts in example diff hunk header

remorses committed:
update react example

remorses committed:
rename showWordHighlights to disableWordHighlights

remorses committed:
chore: remove verbose comments

remorses committed:
refactor: simplify word highlights code

remorses committed:
fix: revert unrelated simplifications to split view forEach loops

remorses committed:
Merge upstream/main into word-highlights

remorses committed:
chore: format Diff.ts

remorses committed:
fix: word highlights scroll with content when scrollX is non-zero

- Added scrollX offset to inline highlight positions in LineNumberRenderable
- Added tests for word highlights with horizontal scrolling in unified and split views

remorses committed:
fix: word highlights respect line wrapping

When text wraps to multiple visual lines, highlights now only appear on the
correct wrapped line(s) by using lineWraps offset to calculate visibility.

remorses committed:
fix: correctly compute column offset for wrapped line highlights

lineWraps contains wrap INDEX (0,1,2...), not column offset.
Now properly computes column offset by summing widths of previous
wrapped segments belonging to the same logical line.

remorses commented:
Also works with text wrapping:

<img width="394" height="256" alt="Screenshot 2025-12-14 at 18 22 01" src="https://github.com/user-attachments/assets/d58062f0-e88a-400e-8c57-0d28d4e600c2" />

remorses committed:
cleanup: remove excessive comments

remorses committed:
test: use inline snapshots with trimmed lines for scrollX tests

remorses committed:
test: remove obsolete external snapshots

remorses committed:
test: remove useless word highlight tests that don't test anything meaningful

remorses committed:
test: remove scrollX tests that can't verify highlight rendering

remorses committed:
chore: format

remorses committed:
Merge branch 'main' into word-highlights

remorses committed:
use diff with words for similarity

remorses committed:
Merge branch 'word-highlights' of https://github.com/remorses/opentui into word-highlights

remorses committed:
increase default lineSimilarityThreshold to 0.5

remorses committed:
fix test: update expected default lineSimilarityThreshold to 0.5

remorses committed:
Merge branch 'main' into word-highlights

Hona reviewed (changes requested):
on pure aesthetics i think it shouldnt be green it should be a more opaque version of the + bg color
thinking about themes and how github does it:

<img width="590" height="191" alt="image" src="https://github.com/user-attachments/assets/2c52c998-4a80-4b86-832c-898ecb77f589" />

you may have already covered that in config right as a new option?
maybe the config you've already done is the way to go

remorses committed:
derive word highlight colors from hunk colors with brighten + opacity increase

remorses committed:
Merge anomalyco/main into word-highlights

Resolved conflicts in Diff.ts and Diff.test.ts, preserving word-level
highlighting feature while incorporating upstream changes.

remorses committed:
Merge remote-tracking branch 'upstream/main' into word-highlights

remorses committed:
refactor: use GitHub Desktop-style word highlights algorithm

- Replace diffWordsWithSpace with prefix/suffix elimination algorithm
- Only highlight when adds.length === removes.length (matching GitHub.com)
- Add MaxIntraLineDiffStringLength (1024 chars) limit
- Remove lineSimilarityThreshold option (not used by GitHub Desktop)
- Update tests for new behavior

remorses committed:
tweak word highlights

remorses committed:
reduce word highlight brightness, use multiplicative alpha

remorses committed:
Merge remote-tracking branch 'upstream/main' into word-highlights

remorses committed:
use RGBA.brighten(1.15) for word highlights

remorses committed:
perf(diff): optimize word highlights + add benchmark

remorses committed:
Merge upstream/main into word-highlights

remorses committed:
fix: suppress MaxListenersExceededWarning in benchmark

remorses committed:
Merge remote-tracking branch 'upstream/main' into word-highlights

remorses committed:
simplify word highlights: remove dead code, unify types, reduce brighten factor

**Dead code removed:**
- RGBA.withAlpha() — added but never called anywhere
- LineNumberRenderable: getInlineHighlights(), setLineInlineHighlights(),
  clearLineInlineHighlights() — only setInlineHighlights and
  clearInlineHighlights are actually used by Diff

**Type unification:**
- Eliminated the InlineHighlight interface from Diff.ts (had startCol,
  endCol, type fields where type was 'added-word' | 'removed-word')
- Now uses LineInlineHighlight (startCol, endCol, bg) directly everywhere
- Introduced lightweight ColumnRange interface for computeInlineHighlights
  return type (just startCol + endCol, no type tag)
- Removed toLineHighlights() mapping method — bg color is now assigned
  at the point where highlights are created in processChangeBlockWithHighlights
- This removes an unnecessary intermediate representation and one layer
  of mapping in both unified and split view code paths

**Brighten factor:**
- Reduced from 1.15 to 1.10 (~10% linear RGB increase instead of ~15%)

remorses committed:
fix: prettier formatting (1.10 -> 1.1)
